src/test/connect-test.c

```diff
@@ -0,0 +1,435 @@
+         */
+
+        addr_4.sin_family = AF_INET;
+        addr_4.sin_port = htobe16(port+1);
+        r = inet_pton(AF_INET, "127.0.0.1", &addr_4.sin_addr);
+        if (r < 0)
+                goto end;
+
+        r = bind(fd_4, (struct sockaddr *) &addr_4, sizeof(addr_4));
+        if (r < 0) {
+                log_error_errno(r, "IPv4 client bind failed: %m");
+                goto end;
+        }
+
+        addr_4.sin_port = htobe16(port);
+
+        r = connect(fd_4, (struct sockaddr *) &addr_4, sizeof(addr_4));
+        if (r == 0)
+                ctx.ipv4 = CONNECTION_SUCCEEDED;
+        else if (r < 0 && errno == EINPROGRESS) {
+                r = sd_event_add_io(e, &event_source, fd_4, EPOLLOUT, conn_handler_4, &ctx);
+                if (r < 0)
+                        goto end;
+        } else {
+                log_error_errno(r, "IPv4 connection failed: %m");
+                goto end;
+        }
+
+        fd_6 = socket(AF_INET6, SOCK_STREAM | SOCK_NONBLOCK | SOCK_CLOEXEC, 0);
+        if (fd_6 < 0) {
+                r = -errno;
+                goto end;
+        }
+
+        r = setsockopt(fd_6, SOL_SOCKET, SO_REUSEADDR, &reuseaddr, sizeof(reuseaddr));
+        if (r < 0)
+                goto end;
+
+        addr_6.sin6_family = AF_INET6;
+        addr_6.sin6_port = htobe16(port+1);
+        r = inet_pton(AF_INET6, "::1", &addr_6.sin6_addr);
+        if (r < 0)
+                goto end;
+
+        r = bind(fd_6, (struct sockaddr *) &addr_6, sizeof(addr_6));
+        if (r < 0) {
+                log_error_errno(r, "IPv6 client bind failed: %m");
+                goto end;
+        }
+
+        addr_6.sin6_port = htobe16(port);
+
+        r = connect(fd_6, (struct sockaddr *) &addr_6, sizeof(addr_6));
+        if (r == 0)
+                ctx.ipv6 = CONNECTION_SUCCEEDED;
+        else if (r < 0 && errno == EINPROGRESS) {
+                r = sd_event_add_io(e, &event_source, fd_6, EPOLLOUT, conn_handler_6, &ctx);
+                if (r < 0)
+                        goto end;
+        } else {
+                log_error_errno(r, "IPv6 connection failed: %m");
+                goto end;
+        }
+
+        if (ctx.ipv4 != CONNECTION_UNDEFINED && ctx.ipv6 != CONNECTION_UNDEFINED) {
+                goto end;
+        }
+
+        /* add the timeout */
+        usec = now(CLOCK_MONOTONIC) + timeout;
+        r = sd_event_add_time(e, &event_source, CLOCK_MONOTONIC, usec, 0, timeout_handler, &ctx);
+        if (r < 0) {
+                log_error_errno(r, "Add timeout error: %m");
+                goto end;
+        }
+
+        /*
+         * Run the event loop -- wait for connections to complete or timeout to
+         * trigger.
+         */
+        r = sd_event_loop(e);
+        if (r < 0) {
+                log_error_errno(r, "Event loop error: %m");
+                goto end;
+        }
+
+end:
+        if (event_source)
+                sd_event_source_unref(event_source);
```

Comment:
please use `_cleanup_(sd_event_source_unrefp)` when declaring the event source, so that this cleanup is done automatically, and we don't need to do it explicitly